File: day11/aoc2025_11.ts
import { readFileSync } from "node:fs";

const graph = new Map<string, string[]>();
for (const line of readFileSync("day11/input11", "utf-8").trim().split("\n")) {
    const [key, rest] = line.split(":");
    graph.set(key, rest.trim().split(" "));
}
const cache = new Map<string, bigint>();
function numPaths(start: string, end: string): bigint {
    const key = `${start},${end}`;
    if (cache.has(key)) return cache.get(key)!;
    const result = start === end
        ? BigInt(1)
        : (graph.get(start) ?? []).reduce((sum, n) => sum + numPaths(n, end), BigInt(0));
    cache.set(key, result);
    return result;
}
console.log(`Part 1: ${numPaths("you", "out")}`);

const part2 = numPaths("svr", "dac") * numPaths("dac", "fft") * numPaths("fft", "out") +
              numPaths("svr", "fft") * numPaths("fft", "dac") * numPaths("dac", "out");
console.log(`Part 2: ${part2}`);

File: day11/aoc2025_11.scala
import scala.collection.mutable

def main(args: Array[String]): Unit = {
    val graph: Map[String, List[String]] = scala.io.Source.fromFile("src/main/scala/input11")
        .getLines().map { line =>
            val parts = line.split(":")
            parts(0) -> parts(1).trim.split(" ").toList
        }.toMap

    val cache = mutable.Map[(String, String), BigInt]()
    def numPaths(start: String, end: String): BigInt = cache.getOrElseUpdate((start, end),
        if (start == end) BigInt(1)
        else graph.getOrElse(start, Nil).map(n => numPaths(n, end)).sum
    )
    println(s"Part 1: ${numPaths("you", "out")}")

    val part2 = numPaths("svr", "dac") * numPaths("dac", "fft") * numPaths("fft", "out") +
                numPaths("svr", "fft") * numPaths("fft", "dac") * numPaths("dac", "out")
    println(s"Part 2: $part2")
}
